Cabal/src/Distribution/Simple/Command.hs

```diff
@@ -631,25 +636,38 @@ commandsRun
   :: CommandUI a
   -> [Command action]
   -> [String]
-  -> CommandParse (a, CommandParse action)
+  -> IO (CommandParse (a, CommandParse action))
 commandsRun globalCommand commands args =
   case commandParseArgs globalCommand True args of
-    CommandHelp help -> CommandHelp help
-    CommandList opts -> CommandList (opts ++ commandNames)
-    CommandErrors errs -> CommandErrors errs
+    CommandDelegate -> pure CommandDelegate
+    CommandHelp help -> pure $ CommandHelp help
+    CommandList opts -> pure $ CommandList (opts ++ commandNames)
+    CommandErrors errs -> pure $ CommandErrors errs
     CommandReadyToGo (mkflags, args') -> case args' of
-      ("help" : cmdArgs) -> handleHelpCommand cmdArgs
+      ("help" : cmdArgs) -> pure $ handleHelpCommand cmdArgs
       (name : cmdArgs) -> case lookupCommand name of
         [Command _ _ action _] ->
-          CommandReadyToGo (flags, action cmdArgs)
-        _ -> CommandReadyToGo (flags, badCommand name)
-      [] -> CommandReadyToGo (flags, noCommand)
+          pure $ CommandReadyToGo (flags, action cmdArgs)
+        _ -> do
+          mCommand <- findExecutable $ "cabal-" <> name
+          case mCommand of
+            Just exec -> callExternal flags exec cmdArgs
+            Nothing -> pure $ CommandReadyToGo (flags, badCommand name)
+      [] -> pure $ CommandReadyToGo (flags, noCommand)
       where
         flags = mkflags (commandDefaultFlags globalCommand)
   where
     lookupCommand cname =
       [ cmd | cmd@(Command cname' _ _ _) <- commands', cname' == cname
       ]
+
+    callExternal :: a -> String -> [String] -> IO (CommandParse (a, CommandParse action))
+    callExternal flags exec cmdArgs = do
+      result <- try $ callProcess exec cmdArgs
```

Comment:
To address one @Kleidukos concern: if we want to run the external command within `cabal exec`, I think we need to change this line:
```diff
- result <- try $ callProcess exec cmdArgs
+ result <- try $ callProcess "cabal" $ ["exec" : exec] ++ cmdArgs
```
… or maybe there is a cleaner way to achieve that?

Reply:
I don't think we want to do that because it shouldn't be assumed the command will run in a context with a `.cabal` file.

Reply:
ok :)